File: sdk/src/main/java/eu/darken/porter/sdk/ShizukuUserServiceCodec.java
package eu.darken.porter.sdk;

import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_COMPONENT;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_DAEMON;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_DEBUGGABLE;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_NO_CREATE;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_PROCESS_NAME;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_REMOVE;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_TAG;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS;
import static eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_VERSION_CODE;

import android.os.Bundle;

import androidx.annotation.NonNull;

import java.util.Objects;

/** How {@link Porter.UserServiceArgs} is written in a Shizuku server's key space. */
final class ShizukuUserServiceCodec {

    private ShizukuUserServiceCodec() {
    }

    /** @param noCreate marks the service "do not start it", as {@code peekUserService} asks for */
    @NonNull
    static Bundle encodeUserService(@NonNull Porter.UserServiceArgs args, boolean noCreate) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(USER_SERVICE_ARG_COMPONENT, args.componentName);
        bundle.putBoolean(USER_SERVICE_ARG_DEBUGGABLE, args.debuggable);
        bundle.putInt(USER_SERVICE_ARG_VERSION_CODE, args.versionCode);
        bundle.putBoolean(USER_SERVICE_ARG_DAEMON, args.daemon);
        bundle.putBoolean(USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS, args.use32BitAppProcess);
        bundle.putString(USER_SERVICE_ARG_PROCESS_NAME,
                Objects.requireNonNull(args.processName, "process name suffix must not be null"));
        if (args.tag != null) {
            bundle.putString(USER_SERVICE_ARG_TAG, args.tag);
        }
        if (noCreate) {
            bundle.putBoolean(USER_SERVICE_ARG_NO_CREATE, true);
        }
        return bundle;
    }

    @NonNull
    static Bundle encodeUserServiceRemoval(@NonNull Porter.UserServiceArgs args, boolean remove) {
        Bundle bundle = new Bundle();
        bundle.putParcelable(USER_SERVICE_ARG_COMPONENT, args.componentName);
        if (args.tag != null) {
            bundle.putString(USER_SERVICE_ARG_TAG, args.tag);
        }
        bundle.putBoolean(USER_SERVICE_ARG_REMOVE, remove);
        return bundle;
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/ShizukuUserServiceCodec.kt
package eu.darken.porter.sdk

import android.os.Bundle
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_COMPONENT
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_DAEMON
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_DEBUGGABLE
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_NO_CREATE
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_PROCESS_NAME
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_REMOVE
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_TAG
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS
import eu.darken.porter.sdk.ShizukuProtocol.USER_SERVICE_ARG_VERSION_CODE

/** How [UserServiceArgs] is written in a Shizuku server's key space. */
internal object ShizukuUserServiceCodec {

    /** @param noCreate marks the service "do not start it", as a peek asks for */
    fun encodeUserService(args: UserServiceArgs, noCreate: Boolean): Bundle = Bundle().apply {
        putParcelable(USER_SERVICE_ARG_COMPONENT, args.componentName)
        putBoolean(USER_SERVICE_ARG_DEBUGGABLE, args.debuggable)
        putInt(USER_SERVICE_ARG_VERSION_CODE, args.version)
        putBoolean(USER_SERVICE_ARG_DAEMON, args.daemon)
        putBoolean(USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS, false)
        putString(USER_SERVICE_ARG_PROCESS_NAME, args.processNameSuffix)
        if (args.tag != null) {
            putString(USER_SERVICE_ARG_TAG, args.tag)
        }
        if (noCreate) {
            putBoolean(USER_SERVICE_ARG_NO_CREATE, true)
        }
    }

    fun encodeUserServiceRemoval(args: UserServiceArgs, remove: Boolean): Bundle = Bundle().apply {
        putParcelable(USER_SERVICE_ARG_COMPONENT, args.componentName)
        if (args.tag != null) {
            putString(USER_SERVICE_ARG_TAG, args.tag)
        }
        putBoolean(USER_SERVICE_ARG_REMOVE, remove)
    }
}
